pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,389 @@
+package kubefed
+
+import (
+	"context"
+	"errors"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	ctlutil "sigs.k8s.io/kubefed/pkg/controller/util"
+	kubefedopt "sigs.k8s.io/kubefed/pkg/kubefedctl/options"
+)
+
+type fakeClientWithUpdateError struct {
+	msg string
+}
+
+var _ client.Client = &fakeClientWithUpdateError{}
+
+func (fake *fakeClientWithUpdateError) Update(ctx context.Context, obj runtime.Object) error {
+	return errors.New(fake.msg)
+}
+
+func (fake *fakeClientWithUpdateError) Get(ctx context.Context, key client.ObjectKey, obj runtime.Object) error {
+	return nil
+}
+
+func (fake *fakeClientWithUpdateError) List(ctx context.Context, opts *client.ListOptions, list runtime.Object) error {
+	return nil
+}
+
+func (fake *fakeClientWithUpdateError) Create(ctx context.Context, obj runtime.Object) error {
+	return nil
+}
+
+func (fake *fakeClientWithUpdateError) Delete(ctx context.Context, obj runtime.Object, opts ...client.DeleteOptionFunc) error {
+	return nil
+}
+
+func (fake *fakeClientWithUpdateError) Status() client.StatusWriter { return nil }
+
+var _ = Describe("Federator", func() {
+	Describe("function Distribute", testDistribute)
+	Describe("helper function createFederatedResource", testCreateFederatedResource)
+})
+
+func testCreateFederatedResource() {
+	var (
+		resource *corev1.Pod
+		expected *unstructured.Unstructured
+		scheme   *runtime.Scheme
+	)
+
+	BeforeEach(func() {
+		resource = newPod("test-pod", "nginx")
+		scheme = runtime.NewScheme()
+	})
+
+	AfterEach(func() {
+		resource = nil
+		expected = nil
+		scheme = nil
+	})
+
+	When("the resource type scheme has NOT been added to the type registry", func() {
+		When("failing to convert to Unstructured", func() {
+			It("should throw an error", func() {
+				var expectedType *UnstructuredConversionError
+				_, err := createFederatedResource(scheme, resource)
+				Expect(err).To(HaveOccurred())
+				Expect(err).To(BeAssignableToTypeOf(expectedType))
+			})
+
+			It("should return a null expected", func() {
+				expected, _ = createFederatedResource(scheme, resource)
+				Expect(expected).To(BeNil())
+			})
+		})
+	})
+
+	When("the resource type scheme has been added to the type registry", func() {
+		BeforeEach(func() {
+			err := corev1.AddToScheme(scheme)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("should not modify the input resource but copy it", func() {
+			expected, _ = createFederatedResource(scheme, resource)
```

Comment:
bit of a nit but this should really be named `actual` as `resource` is really the expected data. Same with the other tests below.